### PR TITLE
Add Top-H decoding (entropy-bounded truncation) as a LogitsWarper for text generation

### PR DESCRIPTION
**Pull request by ErfanBaghaei**
### PR Description

**Add Top-H decoding (entropy-bounded truncation) as a LogitsWarper**

---

#### Summary

This PR adds **Top-H**, a new sampling method for text generation, as a `LogitsWarper` in 🤗 Transformers.

Top-H truncates the next-token distribution by greedily including the most probable tokens until the entropy of the renormalized subset `q` is bounded:

$$
H(q) \leq \alpha \cdot H(p)
$$

where `H(p)` is the entropy of the model’s original distribution and `α ∈ (0,1]` is the user-specified scale.

This provides an **adaptive trade-off between creativity and coherence**: when the model is confident, the subset remains tight; when the model is uncertain, the subset widens—improving diversity without collapsing into noise.

---

#### API

* **New generation config fields**:

  * `top_h: Optional[float] = None`

* **New class**:

  * `TopHLogitsWarper`

Usage with `GenerationConfig`:

```python
from transformers import AutoModelForCausalLM, AutoTokenizer, GenerationConfig

tok = AutoTokenizer.from_pretrained("meta-llama/Llama-3.1-8B")
model = AutoModelForCausalLM.from_pretrained("meta-llama/Llama-3.1-8B")

gen_cfg = GenerationConfig(do_sample=True, temperature=1.3, top_h=0.4)
out = model.generate(**tok("Once upon a time,", return_tensors="pt"), generation_config=gen_cfg, max_new_tokens=50)
print(tok.decode(out[0], skip_special_tokens=True))
```

---

#### Motivation & Results

* **Theoretical foundation**: Top-H is derived from *Entropy-Constrained Maximum Distortion (ECMD)* and its dual *Entropy-Constrained Maximum Mass (ECMM)*. The exact solution is NP-hard; Top-H is a greedy approximation with early termination.
* **Empirical results** (see [paper](https://arxiv.org/abs/2509.02510)):

  * On GSM8K at temperature `T=2`, Top-H improves accuracy by up to **25.63% over min-p** while preserving fluency.
  * On creative writing tasks, Top-H improves diversity without incoherence.
  * Runtime overhead is negligible (**\~0.8%** in benchmarks).
* **Design criteria**: Fully contained in a logits warper, no long-term maintenance burden. ✔️

---

#### Implementation Notes

  *  Vectorized and numerically stable:
  * Uses cumulative sums of `p` and `p log p` for efficiency (O(V log V), same as top-p).
  * Compatible with all other processors/warpers.
---

#### Tests

Added in `tests/generation/test_logits_process.py`:

* In highly peaked distributions, small alpha keeps only the top-1 token.
* Handles zero probability tokens. 
* No-op when `α=1.0`.

---

#### References

* *Top-H Decoding: Adapting the Creativity and Coherence with Bounded Entropy in Text Generation*
  ArXiv: [2509.02510](https://arxiv.org/abs/2509.02510)
* Original implementation: [github.com/ErfanBaghaei/Top-H-Decoding](https://github.com/ErfanBaghaei/Top-H-Decoding)

---

### CODE REVIEW

File: src/transformers/generation/logits_process.py [hunk 1]
```diff
@@ -582,6 +582,122 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> to
...
 
 
+class TopHLogitsWarper(LogitsProcessor):
+    """
+    [`LogitsProcessor`] that implements Top-H sampling, a decoding method which adaptively selects a subset of
+    high-probability tokens based on entropy and cumulative probability constraints.
+
+    This method dynamically determines how many tokens to keep by analyzing the entropy difference of the selected
+    distribution, thereby balancing exploration and exploitation. It ensures that generated text maintains both
+    diversity and coherence.
```
Comment (gante) [inline src/transformers/generation/logits_process.py:591]: Suggestion: add a paragraph with a reference to your paper 🤗 That way, users can quickly find the reference.

File: src/transformers/generation/logits_process.py [hunk 2]
```diff
@@ -582,6 +582,122 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> to
...
+
+    Args:
+        top_h (`float`, *optional*, defaults to 0.4):
+            Scaling coefficient for the entropy-based threshold (`tau`). Must be in the range `(0, 1]`.
```
Comment (gante) [inline src/transformers/generation/logits_process.py]: Let's use no default.

Rationale: `top_h` is an important parameter that we want the user to accurately control. Default values might make input validation harder in the future ("is `0.4` our default, or is it a value explicitly set by the user?")

File: src/transformers/generation/logits_process.py [hunk 3]
```diff
@@ -582,6 +582,122 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> to
...
+
+    >>> inputs = tokenizer("A sequence: 1, 2", return_tensors="pt")
+
+    >>> outputs = model.generate(**inputs, do_sample=True, top_h=0.4)
+    >>> print(tokenizer.batch_decode(outputs, skip_special_tokens=True)[0])
+    A sequence: 1, 2, 3, 4, 5, 6, 7, 8, 9
+    ```
+    """
+
+    def __init__(self, top_h: float = 0.4, filter_value: float = -float("Inf")):
```
Comment (gante) [inline src/transformers/generation/logits_process.py]:
```suggestion
    def __init__(self, top_h: float, filter_value: float = -float("Inf")):
```
(see comment above)

File: src/transformers/generation/logits_process.py [hunk 4]
```diff
@@ -582,6 +582,122 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> to
...
+        super().__init__()
+
+        # input checks
+        if not (0 < top_h <= 1):
+            raise ValueError("alpha must be in the range (0, 1].")
```
Comment (gante) [inline src/transformers/generation/logits_process.py]:
```suggestion
            raise ValueError("`top_h` must be in the range (0, 1].")
```

File: src/transformers/generation/logits_process.py [hunk 5]
```diff
@@ -582,6 +582,122 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> to
...
+        self.top_n = 100
```
Comment (gante) [inline src/transformers/generation/logits_process.py:630]: Let's add a comment here explaning what this variable does, in case some power user wants to tinker with it 🤗

File: src/transformers/generation/logits_process.py [hunk 6]
```diff
@@ -582,6 +582,122 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> to
...
+        self.coef = top_h
```
Comment (gante) [inline src/transformers/generation/logits_process.py]: `coef` is a vague descriptor for this immutable attribute, so let's pick another variable name.

My preference is `self.top_h` but, if you use a different notation in your resources (paper, blog posts, ...), we can stick to your notation 🤗

File: src/transformers/generation/logits_process.py [hunk 7]
```diff
@@ -582,6 +582,122 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> to
...
+        Args:
+            probs (`torch.FloatTensor`):
+                Probability distribution over tokens.
+
+        Return:
+            `torch.FloatTensor`: Scalar entropy value.
+        """
+        probs = probs[probs > 0]
+        probs = probs / torch.sum(probs)
+        return -torch.sum(probs * torch.log2(probs))
```
Comment (gante) [inline src/transformers/generation/logits_process.py]: Is `torch.distributions.Categorical(logits=scores).entropy()` equivalent to this operation? If so, let's use torch's code 

(= let's rely on the functionality of a core depedency as much as possible)

File: src/transformers/generation/logits_process.py [hunk 8]
```diff
@@ -582,6 +582,122 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> to
...
+
+        for b in range(batch_size):
+            # top-k for this example
+            top_probs, top_idx = torch.topk(probs[b], top_n, largest=True, sorted=True)
+
+            # entropy-based threshold tau (computed on the top-k distribution)
+            alpha_sum = top_probs.sum()
+            tau = (self.calculate_entropy(top_probs) - torch.log2(alpha_sum)) * alpha_sum * self.coef
+
+            # grow the kept set until the stopping rule triggers
```
Comment (gante) [inline src/transformers/generation/logits_process.py]: let's briefly describe the stopping rule in this comment

File: src/transformers/generation/logits_process.py [hunk 9]
```diff
@@ -582,6 +582,122 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> to
...
+            sigma = top_probs[0]
+            H = -top_probs[0] * torch.log2(top_probs[0])
```
Comment (gante) [inline src/transformers/generation/logits_process.py]: `H` is not a valid snake case variable name :) let's pick a different (descriptive) variable name

File: src/transformers/generation/logits_process.py [hunk 10]
```diff
@@ -582,6 +582,122 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> to
...
+            chosen = []
+            ind = 0
+            for idx, p in zip(top_idx, top_probs):
```
Comment (gante) [inline src/transformers/generation/logits_process.py]:
```suggestion
            for token_id, token_probability in zip(top_idx, top_probs):
```
(or similar -- the important here is to be very precise in naming variables)

File: src/transformers/generation/logits_process.py [hunk 11]
```diff
@@ -582,6 +582,122 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> to
...
+
+        # compute probabilities
+        scaled_logits = scores
+        probs = torch.softmax(scaled_logits, dim=-1)
+
+        keep_mask = torch.zeros((batch_size, vocab_size), dtype=torch.bool, device=device)
+
+        top_n = min(self.top_n, vocab_size)
+
+        for b in range(batch_size):
```
Comment (gante) [inline src/transformers/generation/logits_process.py]: General comment: as the batch size grows, `for` loops become expensive. Can this loop be vectorized? If yes, that would be a great long-term win 🤗 The same comment apply to the `for` loop below.

If you find these loops hard to vectorize / don't have the bandwidth to do it, please leave a comment like `TODO: vectorize (we have tried X but it didn't work)` so we can pick it up in the future in case it becomes a bottleneck

**Review by gante — COMMENTED** [review]
Thank you for opening the PR! Looks like it's going in the right direction 🔥 The comments I added are minor things, mostly to keep `TopHLogitsWarper` in line with the rest of the library and our coding philosophy 🤗 

Missing: an integration test (with batch size == 1 and batch size > 1). See [here](https://github.com/huggingface/transformers/blob/6254bb4a6811f2a6e87d973b27ed4e878b7ce6f4/tests/generation/test_utils.py#L3048) for an example -- the idea is to hardcode input-output pairs for a specific model, setting the random seed for full reproducibility. Bonus points if you can exactly reproduce input-output pairs that you've seen in your experiments, using a <7B model.

**Comment by gante** [comment]
@ErfanBaghaei ping me when you're happy with the PR, and want a new review :)

**Comment by ErfanBaghaei** [comment]
@gante  Thank you so much for you comments, the PR is now ready for further review.

**Comment by Hellisotherpeople** [comment]
Why did you treat this LLM sampler far different than our PR for P-less decoding: https://github.com/huggingface/transformers/pull/39236

Why doesn't Top-H fall under the same logic that you gave before?

There is a paper showing significant improvement over min_p as well now for P-less decoding!!!! https://arxiv.org/abs/2509.23234

The paper's been accepted to a workshop at NeurIPS so it has been peer reviewed.

**Comment by gante** [comment]
@Hellisotherpeople you're absolutely correct. I've added a comment in the original PR

**Review by gante — COMMENTED** [review]
@ErfanBaghaei Thank you for iterating! The PR looks good to me, minus the issue below 🤗 

- There should be no diff in the `.circleci` and `.github` folders. Looking at the diff, you may need to update `ruff` and run `make fixup`. LMK if you get stuck :)

**Comment by ErfanBaghaei** [comment]
@gante Thank you for the review. I think the problem is resolved now.

**Review by gante — APPROVED** [review]
Thank you for iterating 🤗
